pkg/pillar/containerd/containerd.go

```diff
@@ -713,32 +728,97 @@ func (client *Client) LKTaskPrepare(name, linuxkit string, domSettings *types.Do
+	return newServiceCtx(ctrdSystemServicesNamespace)
+}
+
+// CtrNewUserServicesCtxWithLease returns a new user service containerd context with a 24 hrs lease
+// and a done func to delete the lease and cancel the context after use.
+func (client *Client) CtrNewUserServicesCtxWithLease() (context.Context, context.CancelFunc, error) {
+	return newServiceCtxWithLease(client.ctrdClient, ctrdServicesNamespace)
+}
+
+// CtrNewSystemServicesCtxWithLease returns a new system service containerd context with a 24 hrs lease
+// and a done func to delete the lease and cancel the context after use.
+func (client *Client) CtrNewSystemServicesCtxWithLease() (context.Context, context.CancelFunc, error) {
+	return newServiceCtxWithLease(client.ctrdClient, ctrdSystemServicesNamespace)
+}
+
+// Util methods
+
+// ctrExec starts the executable in a running container and attaches its logging to memlogd
+func (client *Client) ctrExec(ctx context.Context, domainName string, args []string) (string, string, error) {
+	if err := client.verifyCtr(); err != nil {
+		return "", "", fmt.Errorf("ctrExec: exception while verifying ctrd client: %s", err.Error())
+	}
+	ctr, err := client.ctrdClient.LoadContainer(ctx, domainName)
+	if err != nil {
+		return "", "", fmt.Errorf("ctrExec: Exception while loading container: %v", err)
+	}
+
+	spec, err := ctr.Spec(ctx)
+	if err != nil {
+		return "", "", err
+	}
+	task, err := ctr.Task(ctx, nil)
+	if err != nil {
+		return "", "", err
 	}
+
+	pspec := spec.Process
+	pspec.Terminal = true
+	pspec.Args = args
+
+	// plumb the process for I/O
 	var (
-		err  error
-		done func(context.Context) error
+		stdOut bytes.Buffer
+		stdErr bytes.Buffer
 	)
-	newCtrdCtx := namespaces.WithNamespace(context.Background(), ctrdServicesNamespace)
-	newCtrdCtx, done, err = client.ctrdClient.WithLease(newCtrdCtx)
+	cioOpts := []cio.Opt{cio.WithStreams(new(bytes.Buffer), &stdOut, &stdErr), cio.WithFIFODir(fifoDir)}
+	// exec-id for task.Exec can NOT be longer than 71 runes, on top of that it has to match:
+	//   ^[A-Za-z0-9]+(?:[._-](?:[A-Za-z0-9]+))*$:
+	process, err := task.Exec(ctx, fmt.Sprintf("%.50s%.20d", domainName, rand.Int()), pspec, cio.NewCreator(cioOpts...))
 	if err != nil {
-		return nil, nil, fmt.Errorf("CtrCreateCtxWithLease: exception while creating lease: %s", err.Error())
+		return "", "", err
 	}
-	return newCtrdCtx, func() error {
-		if err := done(newCtrdCtx); err != nil {
-			err = fmt.Errorf("CtrCreateCtxWithLease: exception while deleting newCtrdCtx: %s", err.Error())
-			log.Errorf(err.Error())
-			return err
+	defer process.Delete(ctx)
+
+	// prepare an exit code channel
+	statusC, err := process.Wait(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	// finally - run it (asynchronously)
+	if err := process.Start(ctx); err != nil {
+		return "", "", err
+	}
+
+	// block until the process exits or the timer fires
+	timer := time.NewTimer(30 * time.Second)
```

Comment:
Isn't ctrExec used to start the guest (which will run forever)?

Reply:
I just moved this method to the bottom.
@rvs can answer this better I guess.

Reply:
Ah - diffs confused me.